src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.SdkArchiveDiff/Archive.cs

```diff
@@ -0,0 +1,160 @@
+            throw new NotSupportedException("Unsupported archive type");
+    }
+
+    public abstract bool Contains(string relativePath);
+
+    public abstract string[] GetFileNames();
+
+    public abstract string[] GetFileLines(string relativePath);
+
+    public abstract Task<byte[]> GetFileBytesAsync(string relativePath);
+
+    public abstract void Dispose();
+
+    public class TarArchive : Archive
+    {
+        private string _extractedFolder;
+
+        private TarArchive(string extractedFolder)
+        {
+            _extractedFolder = extractedFolder;
+        }
+
+        public static async Task<TarArchive> Create(string path, CancellationToken cancellationToken = default)
+        {
+            var tmpFolder = Directory.CreateTempSubdirectory(nameof(FindArchiveDiffs));
+            using (var gzStream = File.OpenRead (path))
+            using (var gzipStream = new GZipStream (gzStream, CompressionMode.Decompress))
+            {
+                await TarFile.ExtractToDirectoryAsync(gzipStream, tmpFolder.FullName, true, cancellationToken);
+            }
+            return new TarArchive(tmpFolder.FullName);
+        }
+
+        public override bool Contains(string relativePath)
+        {
+            return File.Exists(Path.Combine(_extractedFolder, relativePath));
+        }
+
+        public override string[] GetFileNames()
+        {
+            return Directory.GetFiles(_extractedFolder, "*", SearchOption.AllDirectories).Select(f => f.Substring(_extractedFolder.Length + 1)).ToArray();
+        }
+
+        public override string[] GetFileLines(string relativePath)
+        {
+            return File.ReadAllLines(Path.Combine(_extractedFolder, relativePath));
+        }
+
+        public override Task<byte[]> GetFileBytesAsync(string relativePath)
+        {
+            var filePath = Path.Combine(_extractedFolder, relativePath);
+            if (!File.Exists(filePath))
+                return Task.FromResult<byte[]>([]);
+            return File.ReadAllBytesAsync(Path.Combine(_extractedFolder, relativePath));
+        }
+
+        public override void Dispose()
+        {
+            if (Directory.Exists(_extractedFolder))
+                Directory.Delete(_extractedFolder, true);
+        }
+    }
+
+    public class ZipFileArchive : Archive
+    {
+        private ZipArchive _archive;
+
+        private ZipFileArchive(ZipArchive archive)
+        {
+            _archive = archive;
+        }
+
+        public static new ZipFileArchive Create(string path)
+        {
+            return new ZipFileArchive(new ZipArchive(File.OpenRead(path)));
+        }
+
+        public override bool Contains(string relativePath)
+        {
+            return _archive.GetEntry(relativePath) != null;
+        }
+
+        public override string[] GetFileNames()
+        {
+            return _archive.Entries.Select(e => e.FullName).ToArray();
+        }
+
+        public override string[] GetFileLines(string relativePath)
```

Comment:
`GetFileLines` and `GetFileBytesAsync` are for future use?

Reply:
Yes, I can remove them for now though to avoid dead code.